### PR TITLE
Alerting: Send information about alert rule to data source in headers

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What is this feature?**
This PR adds more headers to the query request:
- `X-Rule-Name` - that contains title of the rule 
- `X-Rule-Folder` - that contains title of the folder where the rule is
- `X-Rule-Version` - contains version of the rule
- `X-Rule-Source` - contains information about what sends the query: scheduler, testing API, or backtesting
- `X-Rule-Type` - type of rule (alerting or recording)

**Why do we need this feature?**
This will allow users better track rule evaluations on the data source end.

**Who is this feature for?**
Loki data source and perhaps every other data source that logs headers

### CODE REVIEW

**Review by cstyan — COMMENTED** [review]
:+1: LGTM, thanks @yuri-tceretian 

In `schedule/alert_rule.go` and `schedule/recording_rule.go` is it also possible to add `X-Rule-Type: alert/recording`? Not familiar with your code base, so I'm not sure if that's the only two entry points for execution of a rule (like if api and backport go through the schedule code eventually).

**Comment by yuri-tceretian** [comment]
> to add X-Rule-Type: alert/recording?

**Comment by cstyan** [comment]
> > to add X-Rule-Type: alert/recording?
> 
> yes, it's there.

:facepalm: not sure how I missed that before, thought for sure I'd checked all the files for it a few times.

:+1: from me but if you want to wait for approval from someone on your team that's fine.

LMK how we could coordinate testing in a dev environment so I can confirm that Loki is able to pick these new fields up :+1:

File: pkg/services/ngalert/models/alert_rule.go
Comment (alexweav) [inline pkg/services/ngalert/models/alert_rule.go:375]: WDYT about just having this be a straight up map of headers, and using the full header name as the key?

The expression pipeline does the same in `expr.Request` even though the thing handling the request might not be contacted over HTTP and might not be out of process at all (TestData datasource) - it uses headers anyway. There's some precedent for it I think

File: pkg/services/ngalert/eval/eval.go
Comment (alexweav) [inline pkg/services/ngalert/eval/eval.go:351]: If we follow other comment about making the condition carry generic headers, this can go there and live next to the other fields. I think it'd be nice to have the full set of headers (org, folder, group, title) defined in one place

**Review by alexweav — APPROVED** [review]
LGTM, one suggestion but it's not blocking

File: pkg/services/ngalert/models/alert_rule.go
Comment (yuri-tceretian) [reply pkg/services/ngalert/models/alert_rule.go:375]: I do not have a strong opinion and can change it if we think that would be clearer\simpler.

 My reasons for having conversion in the intermediate layer between the scheduler and expressions service are the following: 
- Calling it `headers` seems too specific because it exposes implementation details of the service domain (expression engine), which the eval package is supposed to isolate.
- Calling it headers also would imply that I would have to use header-specific syntax like `X-Rule-Uid` for keys, which apart from looking awkward, limits usage of this data only in the headers. However, I think in the case when the query service is a micro-service, this information could be provided downstream in a format other than headers where it could be used for enriching logging context or tracing information. 
- This would also assume that the scheduler (or other places that specify the "header") should properly prepare the map to be compatible with HTTP specifications but if the expression service changes the API, there may be other mechanisms to provide request metadata.

Considering all the above + the fact that this change introduces a new term in the domain of rule evaluators, I decided to make the field name neutral and leave it up to eval package implementation to decide what to do with the information.

File: pkg/tsdb/loki/parse_query.go
Comment (yuri-tceretian) [reply pkg/tsdb/loki/parse_query.go:141]: @svennergr, @cstyan and I had the same impression but our change did not work until after we added it. We do not plan to merge the last two commits in this PR, and open separate changes to have a more elaborate discussion. 
I would appreciate if you could help us figure out why `queryContext.Headers` were not pick up before, and perhaps show place where they're propagated to the http request.

Comment (svennergr) [reply pkg/tsdb/loki/parse_query.go:141]: Ah, right, I see. We use the HTTP client from Grafana's Plugin SDK and those headers are set in this middleware: https://github.com/grafana/grafana-plugin-sdk-go/blob/main/backend/httpclient/custom_headers_middleware.go#L13

However, we don't necessarily create the HTTP client for each request, but rather when Grafana creates an "instance" of the datasource: https://github.com/grafana/grafana/blob/c80b31a0d441fe096169d128bc8eb2810cb77d6a/pkg/tsdb/loki/loki.go#L91-L99

**Pull request merged**
